refactor(example/gwlb_with_vmseries): Refactor Gateway Load Balancer example

sebastianczech opened the pull request:
## Description

PR is a part of #9 and it delivers Gateway Load Balancer in refactored example `gwlb_with_vmseries`.

## Motivation and Context

In #3 all modules and examples are being refactored according to issue description. [PR 372 in archived repository](https://github.com/PaloAltoNetworks/terraform-azurerm-vmseries-modules/pull/372) delivered changes in `gwlb` module, but it didn't contain refactored example with Gateway Load Balancer.

## How Has This Been Tested?

Code was tested by deploying example in the lab.

## Types of changes

- New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes if appropriate.
- [x] All new and existing tests passed.

alperenkose commented on examples/gwlb_with_vmseries/example.tfvars:
```diff
@@ -0,0 +1,281 @@
+    storage_network_security = {
+      vnet_key            = "transit"
+      allowed_subnet_keys = ["management"]
+      allowed_public_ips  = ["134.238.135.14", "134.238.135.140"]
```
same here

sebastianczech replied:
Updated. 

Nevertheless in other examples we have the same IPs: https://github.com/PaloAltoNetworks/terraform-azurerm-swfw-modules/blob/5e8e2660b79d9adef45261abe9ba456dfc64bf6d/examples/dedicated_vmseries/example.tfvars#L25

alperenkose commented on examples/gwlb_with_vmseries/variables.tf:
```diff
@@ -0,0 +1,818 @@
+
+    Please refer to [module documentation](../../modules/loadbalancer/README.md#frontend_ips) for available properties.
+
+    > [!NOTE]
```
need to replace with `**Note** \` ?

sebastianczech replied:
Updated

alperenkose commented on examples/gwlb_with_vmseries/versions.tf:
```diff
@@ -0,0 +1,22 @@
+    random = {
+      source = "hashicorp/random"
+    }
+    http = {
+      source = "hashicorp/http"
+    }
```
I think we don't use this in this example?

sebastianczech replied:
Updated

alperenkose reviewed (commented):
IMO we should keep Common and GWLB examples separate as GWLB is not in the reference architecture design guide yet. We also have links from pan.dev to the repository examples for these designs so merging them with different example.tfvars can cause confusion.

General question; if we keep the GWLB example separate shall we still keep the un-compatible modules in the example code like NATGW or AppGW ? I believe these resources are not possible to integrate with GWLB, and it might bring confusion/doubts to the users trying out the example.

sebastianczech commented on examples/common_vmseries/README.md:
```diff
@@ -513,7 +516,7 @@ Following properties are available:
   Please refer to [module documentation](../../modules/loadbalancer/README.md#frontend_ips) for available properties.
 
-  > [!NOTE] 
+  > [!NOTE]
```
I was not touching load balancers in common design - I added only GWLB to common design. 
The same we have in other reference architecture examples e.g. https://github.com/PaloAltoNetworks/terraform-azurerm-swfw-modules/blob/5e8e2660b79d9adef45261abe9ba456dfc64bf6d/examples/dedicated_vmseries/variables.tf#L222

I propose to update LB settings in all examples in separate PR, ok ?

Here VS code automatically remove blank characters at the end of the line, so that's why you see changed file.

sebastianczech replied:
I removed all changes from `examples/common_vmseries/README.md`

sebastianczech commented on examples/gwlb_with_vmseries/example.tfvars:
```diff
@@ -0,0 +1,281 @@
+            direction                  = "Inbound"
+            access                     = "Allow"
+            protocol                   = "Tcp"
+            source_address_prefixes    = ["134.238.135.14", "134.238.135.140"]
```
Updated. 

Nevertheless in other examples we have the same IPs: https://github.com/PaloAltoNetworks/terraform-azurerm-swfw-modules/blob/5e8e2660b79d9adef45261abe9ba456dfc64bf6d/examples/dedicated_vmseries/example.tfvars#L25

acelebanski commented:
So summing up today's discussion during the planning meeting, we decided not to share common codebase with GWLB example, it will remain separate. 

We will also transform the **test_infrastructure** example into a module and then integrate it with all examples but for now (in scope of this PR), I think the test infra can remain a part of the GWLB example.

sebastianczech commented on examples/common_vmseries/variables.tf:
```diff
@@ -219,7 +219,7 @@ variable "load_balancers" {
     Please refer to [module documentation](../../modules/loadbalancer/README.md#frontend_ips) for available properties.
 
-    > [!NOTE] 
+    > [!NOTE]
```
I removed all changes from `examples/common_vmseries`

sebastianczech commented:
thanks @alperenkose for review, I updated the code.

@acelebanski I removed changes from common design and I left refactored example for GWLB together test VMs. 

PR is ready to being reviewed.

acelebanski commented on examples/gwlb_with_vmseries/main.tf:
```diff
@@ -0,0 +1,338 @@
+  health_probe = try(each.value.health_probe, null)
+  lb_rule      = try(each.value.lb_rule, null)
+
+  zones = var.enable_zones ? try(each.value.zones, null) : null
```
What does `each.value.zones` refer to if there's no `zones` property inside `gateway_load_balancers` variable?

sebastianczech replied:
Thanks, I missed that. Add.

acelebanski replied:
I think that now variable `enable_zones` is redundant. You can eliminate it completely.

sebastianczech replied:
Well, this I done by purpose - we have `enable_zones` for other examples. I wanted to have the same parts (as much as possible) in comparison to common example e.g. I removed NAT GW and application GW, but I added GWBL and test VMs. Should I remove `enable_zones` from all examples ?

sebastianczech replied:
I remove `enable_zones` from GWLB example. In this PR I'm not touching other examples, ok ?

acelebanski replied:
Sure, thanks!

acelebanski commented on examples/gwlb_with_vmseries/templates/bootstrap-gwlb.tftpl:
Is it required to put the entire VM-Series config in a template when in conjunction with GWLB?

sebastianczech replied:
If we want to have working example (without any additional configuration on VM-Series) after `terraform apply`, then it's required to have the entire VM-Series config in a template.

In previous version of the example (before refactor) we had it.

Nevertheless we can remove it entirely, if we don't want to have everything configured - should I remove it ?

acelebanski replied:
I was asking more out of curiosity. We do the same thing with **dedicated_vmseries** example so I think that's OK to have it. Customer may use it or not. 👍
